Allow Arbitrary Initial Pose of Model in World Frame

liangfok opened the pull request:
Resolves https://github.com/RobotLocomotion/drake/issues/2003
Resolves #2058

# Key Changes

* Modified the semantics of `RigidBodyTree:addRobotFromFile()` to allow arbitrary positioning of a model's world frame within Drake's world frame.
* Added method `RigidBodyTree::AddFloatingJoint()` and modified `RigidBodyTreeURDF` and `RigidBodySDF` to use it. This consolidates what would otherwise be duplicate code between the URDF and SDF parsers.
* Added method `RigidBodyTree::AddRigidBody()` and modified `RigidBodyTreeURDF` and `RigidBodySDF` to use it.
* Added unit tests `load_model_test` and `rigid_body_tree_test` to verify the correct implementation of the above-mentioned changes.

# Execution of New Unit Tests

```
$ cd [drake root]/drake/pod-build
$ ctest -R load_model_test
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2053) &emsp; Multiple assignees:&emsp;<img alt="@RussTedrake" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/6442292?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/RussTedrake">RussTedrake</a>,&emsp;<img alt="@david-german-tri" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/17437069?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/david-german-tri">david-german-tri</a>
<!-- Reviewable:end -->

liangfok commented:
@david-german-tri, once this passes the CI test suite, can you perform a first-round review?

david-german-tri commented:
CI failure on win32 appears to be genuine.

---

Review status: 0 of 13 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-:-KF4Wd0UEohX9hxslFbD:1353956488)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I agree is genuine. Here is the error:

```
C:\projects\workspace\experimental\e052ff41\drake\systems\plants\test\load_urdf_test.cc(46): error C2718: 'const Eigen::Matrix<double,4,4,0,4,4>': actual parameter with requested alignment of 16 won't be aligned 
```

Here is the problematic code:
```
  EXPECT_EQ(link1_body->getJoint().getTransformToParentBody().matrix(),
            Eigen::Isometry3d::Identity().matrix());
```

It appears that I cannot pass an `Eigen::Matrix` into Google Test's `EXPECT_EQ()` macro without disabling the alignment requirement. Since I don't want to disable alignment for performance reasons, I'm going to try changing the code to be:

```
EXPECT_TRUE(link1_body->getJoint().getTransformToParentBody().matrix() == 
            Eigen::Isometry3d::Identity().matrix());
```

---

Review status: 0 of 13 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-:-KF4_9dJqIXgMcaPwzvE:-1570248326)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Here are some design comments; I've skipped style comments for now.

---

Reviewed 2 of 13 files at r1.
Review status: 1 of 13 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.h, line 740 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KF4__hZz1x-EeiRMyeY:-KF4__hZz1x-EeiRMyeZ:1865197212) ([raw file](https://github.com/robotlocomotion/drake/blob/712ec0a9c6fa502ad25355ef41252059eb480d3f/drake/systems/plants/RigidBodyTree.h#L740)):*
This sentence is confusing because the RigidBodyTree parameter is itself named `model`.

---

*[drake/systems/plants/RigidBodyTree.h, line 756 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KF4_-Yk6HvAaFIlnswe:-KF4_-Yk6HvAaFIlnswf:1370042556) ([raw file](https://github.com/robotlocomotion/drake/blob/712ec0a9c6fa502ad25355ef41252059eb480d3f/drake/systems/plants/RigidBodyTree.h#L756)):*
Why did you make this a static method with an explicit `model` parameter, instead of an ordinary class method?

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 526 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KF4c-wJLNTKslOZXqsN:-KF4c-wJLNTKslOZXqsO:-881218838) ([raw file](https://github.com/robotlocomotion/drake/blob/712ec0a9c6fa502ad25355ef41252059eb480d3f/drake/systems/plants/RigidBodyTreeSDF.cpp#L526)):*
What does "API-based shifting" mean?

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 578 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KF4cJ4x0QAwt5ZSPgE6:-KF4cJ4x0QAwt5ZSPgE7:-1023683774) ([raw file](https://github.com/robotlocomotion/drake/blob/712ec0a9c6fa502ad25355ef41252059eb480d3f/drake/systems/plants/RigidBodyTreeSDF.cpp#L578)):*
To make sure I understand this: For every`< model>` in the SDF, we will loop over every body in the RigidBodyTree to check for parentless nodes.  In the case where there are many models M with many links L, this is O(M*L), even though each link will actually get a floating joint at most once.  Is that right?

If so...the quadratic-ish performance may not be a big deal in practice, but it seems like you could fix it, and also make the code clearer, by adding the floating joints as the links are parsed.  What do you think?

---

*[drake/systems/plants/test/load_sdf_test.cc, line 1 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KF4hekS7F9buHORGRgT:-KF4hekT0M7sb06WIWl3:1865176552) ([raw file](https://github.com/robotlocomotion/drake/blob/712ec0a9c6fa502ad25355ef41252059eb480d3f/drake/systems/plants/test/load_sdf_test.cc#L1)):*
These two test files are basically identical aside from the filenames, right?  Could you combine them using a TEST_P to eliminate the code duplication and make sure they stay in sync going forward?

https://github.com/google/googletest/blob/master/googletest/docs/AdvancedGuide.md#how-to-write-value-parameterized-tests

---

*[drake/systems/plants/test/load_sdf_test.cc, line 39 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KF4gpzPltdJGcRkX9lG:-KF4gpzPltdJGcRkX9lH:2059945465) ([raw file](https://github.com/robotlocomotion/drake/blob/712ec0a9c6fa502ad25355ef41252059eb480d3f/drake/systems/plants/test/load_sdf_test.cc#L39)):*
You can just say EXPECT_EQ(link1_body->linkname, "link1")

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-:-KF4fiFuapdr6kRR1MlF:1722751119)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Thanks!

---

Review status: 0 of 13 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.h, line 740 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KF4__hZz1x-EeiRMyeY:-KF5IizhvjoLMW6TtP81:-923758060) ([raw file](https://github.com/robotlocomotion/drake/blob/712ec0a9c6fa502ad25355ef41252059eb480d3f/drake/systems/plants/RigidBodyTree.h#L740)):*
I've reworded the comment.

---

*[drake/systems/plants/RigidBodyTree.h, line 756 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KF4_-Yk6HvAaFIlnswe:-KF5JI_6wa7P-7oIVa4n:855679332) ([raw file](https://github.com/robotlocomotion/drake/blob/712ec0a9c6fa502ad25355ef41252059eb480d3f/drake/systems/plants/RigidBodyTree.h#L756)):*
I was following the design of the methods that call this method in `RigidBodyTreeSDF.cpp`:

```
void parseModel(RigidBodyTree* model, XMLElement* node,
                const PackageMap& package_map, const string& root_dir,
                const DrakeJoint::FloatingBaseType floating_base_type,
                std::shared_ptr<RigidBodyFrame> weld_to_frame)
```

and `RigidBodyTreeURDF.cpp`:

```
void parseRobot(RigidBodyTree* model, XMLElement* node,
                const PackageMap& package_map, const string& root_dir,
                const DrakeJoint::FloatingBaseType floating_base_type,
                std::shared_ptr<RigidBodyFrame> weld_to_frame = nullptr)
```

I suppose since it's possible and simpler to make `AddFloatingJoint()` an ordinary class method, I will modify it to be one.

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 526 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KF4c-wJLNTKslOZXqsN:-KF4lIeB9oDPV7OdbpdV:-1851366215) ([raw file](https://github.com/robotlocomotion/drake/blob/712ec0a9c6fa502ad25355ef41252059eb480d3f/drake/systems/plants/RigidBodyTreeSDF.cpp#L526)):*
I was referring to the `weld_to_joint` parameter

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 578 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KF4cJ4x0QAwt5ZSPgE6:-KF4l352WLKFpDFtCrI4:-116552098) ([raw file](https://github.com/robotlocomotion/drake/blob/712ec0a9c6fa502ad25355ef41252059eb480d3f/drake/systems/plants/RigidBodyTreeSDF.cpp#L578)):*
You're right. The current implementation of `AddFloatingJoint()` iterates over every link in the rigid body tree, even those that were already part of the tree, and an optimization would be to avoid having to consider every node in the tree each time a model is added to the tree.

Unfortunately, it is not possible to add floating joints while parsing the links since we cannot tell which links will remain parent-less until after _all_ links and _all_ joints within a model are parsed.

One possibility is to pre-parse the model by analyzing every link and joint in it to identify which ones are the root links. This will incur a time complexity of O(L).

Another possibility is to maintain a list of newly added links as they are being added, and to remove links from this list as joints are being added. Once all links and joints are added, anything remaining in this list would be parent-less and thus require floating joints.  `AddFloatingJoint()` could then iterate over this list. I believe the time complexity of this would be O(log(L)) since searches will need to be performed over the list to remove links.

Since this is optimization and not directly related to this PR, I would like to kick this optimization down the road and address it in a future PR. Here is a newly created issue to track it: https://github.com/RobotLocomotion/drake/issues/2058

---

*[drake/systems/plants/test/load_sdf_test.cc, line 1 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KF4hekS7F9buHORGRgT:-KF5hobPtkjLYRAbYTBp:375385027) ([raw file](https://github.com/robotlocomotion/drake/blob/712ec0a9c6fa502ad25355ef41252059eb480d3f/drake/systems/plants/test/load_sdf_test.cc#L1)):*
Good point. I've updated the unit test to use TEST_P. Thanks!

Note that the new combined unit test is in load_model_test.cc.

---

*[drake/systems/plants/test/load_sdf_test.cc, line 39 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KF4gpzPltdJGcRkX9lG:-KF5_sgD18hLyaUXRnFF:-795674665) ([raw file](https://github.com/robotlocomotion/drake/blob/712ec0a9c6fa502ad25355ef41252059eb480d3f/drake/systems/plants/test/load_sdf_test.cc#L39)):*
Fixed. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-:-KF4irvzX57DpT2MB-oC:-1834867277)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@david-german-tri: Thanks for the design comments. PTAL.

---

Review status: 0 of 13 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-:-KF5v2_o9yVK-zkNHU1V:-1135932458)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 13 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 526 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KF4c-wJLNTKslOZXqsN:-KF5wDdywBVAkfZvOKrb:-1487817258) ([raw file](https://github.com/robotlocomotion/drake/blob/712ec0a9c6fa502ad25355ef41252059eb480d3f/drake/systems/plants/RigidBodyTreeSDF.cpp#L526)):*
Sorry, forgot to finish my reply. I updated the Doxygen comment to be more clear. PTAL. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

liangfok commented:
CI test suite's Win64/32 tests identified a strange problem:

```
load_model_test.obj : error LNK2019: unresolved external symbol "class testing::internal::Mutex testing::internal::g_linked_ptr_mutex"
```

From [this](https://groups.google.com/forum/#!topic/googlemock/Ejyq8M736Ec) thread, it appears that I need to " include the GTEST_LINKED_AS_SHARED_LIBRARY=1 to the dll code". Unfortunately, I'm not sure what that means.

liangfok commented:
@david-german-tri: I believe I encountered a Windows flake. See [this console output.](https://drake-jenkins.csail.mit.edu/job/experimental/815/compiler=msvc-32,label=windows/console).

david-german-tri commented:
See #2075.

david-german-tri commented:
jenkins: retest this please

liangfok commented:
@david-german-tri, this branch is ready for another round of feature review.

david-german-tri commented:
Nice improvements!  Here are a handful of style things and a couple more design questions.

---

Reviewed 1 of 8 files at r3, 1 of 3 files at r8.
Review status: 2 of 13 files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 1868 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEonbi2dwDeHLHEPI9:-KFEonbi2dwDeHLHEPIA:653034585) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/RigidBodyTree.cpp#L1868)):*
weld_to_frame->name == "world"

---

*[drake/systems/plants/RigidBodyTree.cpp, line 1888 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEp7ebKbG2doJftR30:-KFEp7ebKbG2doJftR31:-1979728794) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/RigidBodyTree.cpp#L1888)):*
"non-world link" (I didn't understand this comment until I read the header file.  The extra word really helps!)

---

*[drake/systems/plants/RigidBodyTree.h, line 745 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEptdOsDYXOLRsOcDh:-KFEptdOsDYXOLRsOcDi:918741766) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/RigidBodyTree.h#L745)):*
This is a particularly awesome comment.  Thanks!

---

*[drake/systems/plants/RigidBodyTree.h, line 754 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEraV_vkjMv9DYUNiu:-KFEraVbNvOTO0B3M5Bu:56629679) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/RigidBodyTree.h#L754)):*
Should this be called AddFloatingJoints?

---

*[drake/systems/plants/RigidBodyTree.h, line 754 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEvPh6Jv1LLJWd8emo:-KFEvPh6Jv1LLJWd8emp:1104930301) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/RigidBodyTree.h#L754)):*
The regression tests are awesome, but could you please also add a unit test for this function?  Doesn't need to be much, but it would be nice to cover the multi-model case.  Something like: add two links, call AddFloatingJoint with a weld_to_frame, add two more links, call AddFloatingJoint without a weld_two_frame, check that everything is welded in the right place.

Ulterior motive: this will also get a unit test file for RigidBodyTree started, which is desperately needed.

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 578 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KF4cJ4x0QAwt5ZSPgE6:-KFErnJzrZz4kRtcmq8_:96875526) ([raw file](https://github.com/robotlocomotion/drake/blob/712ec0a9c6fa502ad25355ef41252059eb480d3f/drake/systems/plants/RigidBodyTreeSDF.cpp#L578)):*
Thanks for the explanation!  I'm primarily concerned about clarity here, not efficiency.  Right now, the reason this code is correct is pretty subtle.  I had to read it three times to convince myself you wouldn't wind up welding links from model 1 to the frame of model 2.  I'm worried that would be easy for someone else to break.

So, I'm going to push a little bit more: Why not just collect a list of all the body indices in the model, pass the list into AddFloatingJoint, and iterate over that instead of all the bodies?

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 530 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEqBY_NGXdomSE-04D:-KFEqBY_NGXdomSE-04E:-795960519) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/RigidBodyTreeSDF.cpp#L530)):*
ROS workspace?

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 585 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEubSq8FHhN7YMfJIn:-KFEubSq8FHhN7YMfJIo:-1357200196) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/RigidBodyTreeURDF.cpp#L585)):*
joints that connect?

---

*[drake/systems/plants/test/load_model_test.cc, line 17 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEvCBXejUm3T_Ru0uh:-KFEvCBXejUm3T_Ru0ui:425719327) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/test/load_model_test.cc#L17)):*
A comment documenting the param would be good.

---

*[drake/systems/plants/test/load_model_test.cc, line 18 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEwnVgkprR6kHNbCt6:-KFEwnVgkprR6kHNbCt7:1061336020) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/test/load_model_test.cc#L18)):*
nit: protected

---

*[drake/systems/plants/test/load_model_test.cc, line 50 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEuqgZ-IDOVvO5jPtH:-KFEuqgZ-IDOVvO5jPtI:2091587157) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/test/load_model_test.cc#L50)):*
I suspect EXPECT_EQ will work here too, but I'm not certain.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-:-KFEwsa0v46Xtb_38hwP:-1102595194)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 1 of 14 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 1868 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEonbi2dwDeHLHEPI9:-KFFYZskqRx6WILY4IBE:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/RigidBodyTree.cpp#L1868)):*
Done.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 1888 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEp7ebKbG2doJftR30:-KFFYpHa-XPb8TcPwl0K:35456933) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/RigidBodyTree.cpp#L1888)):*
Oops. You're right. I accidentally omitted that word. Thanks for catching this.

---

*[drake/systems/plants/RigidBodyTree.h, line 754 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEraV_vkjMv9DYUNiu:-KFF_7dCNn84fxku6BLT:960124142) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/RigidBodyTree.h#L754)):*
Yes. Changed.

---

*[drake/systems/plants/RigidBodyTree.h, line 754 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEvPh6Jv1LLJWd8emo:-KFG2bctD_TtFwfGSZ2q:359985724) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/RigidBodyTree.h#L754)):*
Good idea. I added a unit test, see rigid_body_tree_test.cc.

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 578 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KF4cJ4x0QAwt5ZSPgE6:-KFG8i3clqMdZSDzwoeL:-1673201522) ([raw file](https://github.com/robotlocomotion/drake/blob/712ec0a9c6fa502ad25355ef41252059eb480d3f/drake/systems/plants/RigidBodyTreeSDF.cpp#L578)):*
I'd prefer to not burden the user of `RigidBodyTree` with the task of creating a list of every link in a newly added model. Perhaps the best design that will not incur additional work on the user is to maintain a cache within `RigidBodyTree` containing pointers to every rigid body that was added since the last time `RigidBodyTree::AddFloatingJoints()` was called. Then, modify `RigidBodyTree:AddFloatingJoints()` to search over the links in the cache searching for parent-less links and attaching floating joints to them. Doing this will require making `RigidBodyTree::bodies` protected, and adding a `RigidBodyTree::addRigidBody()` method through which new rigid bodies can be added both to the `bodies` vector and the aforementioned cache. What do you think about this proposal?  Should I include this change in this PR or create an issue to handle it in a future PR?

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 530 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEqBY_NGXdomSE-04D:-KFG36qQkT2STsFqMHaf:987855027) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/RigidBodyTreeSDF.cpp#L530)):*
Yes, ROS workspace. I verified this with @avalenzu. See: https://github.com/RobotLocomotion/drake/commit/9dbc390b0543e6644e8c96418a0454aa1e4b2451#commitcomment-16956769

---

*[drake/systems/plants/RigidBodyTreeURDF.cpp, line 585 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEubSq8FHhN7YMfJIn:-KFGA-W9X1eiMz6Tv2Zk:-775413954) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/RigidBodyTreeURDF.cpp#L585)):*
It could be singular or plural. Thus, I will use "joint(s)"

---

*[drake/systems/plants/test/load_model_test.cc, line 17 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEvCBXejUm3T_Ru0uh:-KFGBRiiE3LV35YHMY6L:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/test/load_model_test.cc#L17)):*
Done.

---

*[drake/systems/plants/test/load_model_test.cc, line 18 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEwnVgkprR6kHNbCt6:-KFGBSp8uOLU0HSQmC6h:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/test/load_model_test.cc#L18)):*
Done.

---

*[drake/systems/plants/test/load_model_test.cc, line 50 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEuqgZ-IDOVvO5jPtH:-KFGBUy3FMGM9PfXTINE:-541204853) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/test/load_model_test.cc#L50)):*
I originally tried using EXPECT_EQ but failed because of an Eigen alignment problem. See: https://github.com/RobotLocomotion/drake/pull/2053#issuecomment-208353902

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Reviewed 6 of 14 files at r1, 1 of 8 files at r3, 7 of 7 files at r11.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 578 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KF4cJ4x0QAwt5ZSPgE6:-KFGG1DA1qiNL-sqWelA:1209540101) ([raw file](https://github.com/robotlocomotion/drake/blob/712ec0a9c6fa502ad25355ef41252059eb480d3f/drake/systems/plants/RigidBodyTreeSDF.cpp#L578)):*
As discussed f2f - you're parsing in the links a few lines above here, so it's pretty unintrusive to just hang onto pointers/indices in the body of this function, and that makes it crystal clear which links are getting welded.

---

*[drake/systems/plants/test/rigid_body_tree_test.cc, line 36 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFGIH_Xo1FB0t2R7nFF:-KFGIH_YeDgWlwIvOjSq:-824738587) ([raw file](https://github.com/robotlocomotion/drake/blob/23331c618df04db961b2747f55232930461727bd/drake/systems/plants/test/rigid_body_tree_test.cc#L36)):*
nit: does the bN part of this notation buy us anything?

---

*[drake/systems/plants/test/rigid_body_tree_test.cc, line 107 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFGFSBjXm7pBDCMFU30:-KFGFSBkP80yImB49nz7:-1581967562) ([raw file](https://github.com/robotlocomotion/drake/blob/23331c618df04db961b2747f55232930461727bd/drake/systems/plants/test/rigid_body_tree_test.cc#L107)):*
I'd suggest moving all of the EXPECT statements to the end, to prove that subsequent welds didn't overwrite any of the parent joints.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 8 of 14 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTreeSDF.cpp, line 578 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KF4cJ4x0QAwt5ZSPgE6:-KFGUMVzjUw7R0RyurCR:-1326275650) ([raw file](https://github.com/robotlocomotion/drake/blob/712ec0a9c6fa502ad25355ef41252059eb480d3f/drake/systems/plants/RigidBodyTreeSDF.cpp#L578)):*
Done. Note that to consolidate the places where the body_index is computed, I added `RigidBodyTree::AddRigidBody()`.

---

*[drake/systems/plants/test/rigid_body_tree_test.cc, line 36 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFGIH_Xo1FB0t2R7nFF:-KFGUkmDJlKibQi760A_:123159778) ([raw file](https://github.com/robotlocomotion/drake/blob/23331c618df04db961b2747f55232930461727bd/drake/systems/plants/test/rigid_body_tree_test.cc#L36)):*
Not presently since each robot only has one link. I think it would best to leave them since we are planning on beefing up the rigid body tree's unit test to include robots with more than a single link.

---

*[drake/systems/plants/test/rigid_body_tree_test.cc, line 107 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFGFSBjXm7pBDCMFU30:-KFGUjPgkO-bN3sBNEIv:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/23331c618df04db961b2747f55232930461727bd/drake/systems/plants/test/rigid_body_tree_test.cc#L107)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 8 of 14 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/test/rigid_body_tree_test.cc, line 140 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFGx3niO_9Yh9ieZKg0:-KFGx3niO_9Yh9ieZKg1:-1304770838) ([raw file](https://github.com/robotlocomotion/drake/blob/a99b985750db5ef1c26e4b5d6cfcc2e07feb2997/drake/systems/plants/test/rigid_body_tree_test.cc#L140)):*
nit: you could replace link_indicies with an initializer list: {r2b1->body_index} (and analogously below)

if you don't, another nit: should be link_indices

---

*[drake/systems/plants/test/rigid_body_tree_test.cc, line 162 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFGxHO5YL8GdNV9jBFe:-KFGxHO5YL8GdNV9jBFf:-1252042790) ([raw file](https://github.com/robotlocomotion/drake/blob/a99b985750db5ef1c26e4b5d6cfcc2e07feb2997/drake/systems/plants/test/rigid_body_tree_test.cc#L162)):*
Might be interesting to add two at once here.  Up to you.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-:-KFGxeIOzUC60VuicxGI:1417088721)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Reviewed 5 of 6 files at r12.
Review status: 13 of 14 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 13 of 14 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/test/rigid_body_tree_test.cc, line 140 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFGx3niO_9Yh9ieZKg0:-KFH5U29idYbEd9szcwk:-294609872) ([raw file](https://github.com/robotlocomotion/drake/blob/a99b985750db5ef1c26e4b5d6cfcc2e07feb2997/drake/systems/plants/test/rigid_body_tree_test.cc#L140)):*
Wow. I didn't even know this was valid C++. It highlights one reason why parameters that _can_ be `const` should be declared `const`. Thanks. TIL.

---

*[drake/systems/plants/test/rigid_body_tree_test.cc, line 162 \[r12\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFGxHO5YL8GdNV9jBFe:-KFH6rS5h-R8vSAcz7VI:1185480827) ([raw file](https://github.com/robotlocomotion/drake/blob/a99b985750db5ef1c26e4b5d6cfcc2e07feb2997/drake/systems/plants/test/rigid_body_tree_test.cc#L162)):*
Done. Added two floating joints in one go.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@david-german-tri: Thanks for the thorough code review. I shall now assign this PR to @RussTedrake for platform review.

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 6 files at r12, 1 of 3 files at r13, 2 of 2 files at r14.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/test/load_model_test.cc, line 221 \[r14\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFPah6jAQdVBOjaq8I1:-KFPah6kEl-_7NWWfgr8:-927446528) ([raw file](https://github.com/robotlocomotion/drake/blob/09d1106b7560212ed09c9f96a49498622298f75c/drake/systems/plants/test/load_model_test.cc#L221)):*
Nice comment!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-:-KFPb2bvINIzvefgh-hx:1417088721)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
+@RussTedrake 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-:-KFPbZF0haMHpC7Sw4nw:-755227487)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
+@eriksobel-tri 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-:-KFPcDfHppdBenPHEBpe:-136465282)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
+@david-german-tri 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-:-KFPcfg7-9uyRGR3_17A:1609708521)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
+@lucy-tri 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-:-KFPe_Ra1caocD4gvOf7:449547561)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
-@ErikSobel-TRI -@Lucy-tri thanks for appearing in the demo!

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-:-KFPv47WNl-jjrLc1L0v:4059361)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 1849 \[r14\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFPxO_NlgzbF0iyEe_c:-KFPxO_OP5z5DJ2FG12D:7935480) ([raw file](https://github.com/robotlocomotion/drake/blob/09d1106b7560212ed09c9f96a49498622298f75c/drake/systems/plants/RigidBodyTree.cpp#L1849)):*
This follows the Google style guide. However the other option  would be to use all lower case as so `add_rigid_body`. The spirit being that cheap to call methods are named all lower case making *expensive to call methods* easily detectable by their upper case signature.  From [Google's style guide](https://google.github.io/styleguide/cppguide.html#Function_Names): 
> Functions that are very cheap to call may instead follow the style for variable names (all lower-case, with underscores between words). The rule of thumb is that such a function should be so cheap that you normally wouldn't bother caching its return value when calling it in a loop. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
Glad to learn about it!
Butterflies!

---

Reviewed 6 of 14 files at r1, 1 of 8 files at r3, 2 of 7 files at r11, 3 of 6 files at r12, 1 of 3 files at r13, 2 of 2 files at r14.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-:-KFQ1cJkRUcYFup0rkIy:129980613)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 9 of 14 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 1849 \[r14\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFPxO_NlgzbF0iyEe_c:-KFQcnsSUzF6OSBUyDUo:1166733888) ([raw file](https://github.com/robotlocomotion/drake/blob/09d1106b7560212ed09c9f96a49498622298f75c/drake/systems/plants/RigidBodyTree.cpp#L1849)):*
I agree and changed the method name.  Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 8 of 14 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.h, line 767 \[r15\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFQkOEAKnvn4XEzzhcd:-KFQkOEAKnvn4XEzzhce:-444029367) ([raw file](https://github.com/robotlocomotion/drake/blob/32684757f7e91a5599e8eb45ba840375216552f3/drake/systems/plants/RigidBodyTree.h#L767)):*
If this method does something "simple" (meaning not computationally intensive), should we also use the lower case + underscores naming convention here also?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 8 of 14 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.h, line 767 \[r15\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFQkOEAKnvn4XEzzhcd:-KFQpE_ADEqUU3qF2Nfv:-1174772496) ([raw file](https://github.com/robotlocomotion/drake/blob/32684757f7e91a5599e8eb45ba840375216552f3/drake/systems/plants/RigidBodyTree.h#L767)):*
I'd say definitely not - it's 50+ lines of code and O(N) in the link_indices.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

liangfok commented:
*[drake/systems/plants/RigidBodyTree.h, line 767 \[r15\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFQkOEAKnvn4XEzzhcd:-KFQptAB6rcyaLlQWR2s:-1963827052) ([raw file](https://github.com/robotlocomotion/drake/blob/32684757f7e91a5599e8eb45ba840375216552f3/drake/systems/plants/RigidBodyTree.h#L767)):*
OK I will keep it as is then. We should clarify the style guide to indicate what methods are considered simple enough for the lower-case-under-score naming convention. Perhaps only methods that are that are O(1) should be considered "simple"?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 8 of 14 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/plants/RigidBodyTree.h, line 767 \[r15\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFQkOEAKnvn4XEzzhcd:-KFQr7A5vw7IdRDuyz8s:-990398595) ([raw file](https://github.com/robotlocomotion/drake/blob/32684757f7e91a5599e8eb45ba840375216552f3/drake/systems/plants/RigidBodyTree.h#L767)):*
My personal rule of thumb is O(1) in all dimensions, and five lines of code or fewer.  I'm OK with formalizing that if you like it and think it adds value, but I'm also OK with a more ad hoc approach: if anyone expresses doubt, don't use an underscore name.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

liangfok commented:
*[drake/systems/plants/RigidBodyTree.h, line 767 \[r15\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFQkOEAKnvn4XEzzhcd:-KFQuUAqZ8OJifjljFuT:1553735460) ([raw file](https://github.com/robotlocomotion/drake/blob/32684757f7e91a5599e8eb45ba840375216552f3/drake/systems/plants/RigidBodyTree.h#L767)):*
What does everyone think about eliminating the options of lower-case-with-under-score method names and standardizing on CamelCase for method names? This would result in less confusion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 8 of 14 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/plants/RigidBodyTree.h, line 767 \[r15\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFQkOEAKnvn4XEzzhcd:-KFVmmM2ckBNhtqYZ9-Q:1146768292) ([raw file](https://github.com/robotlocomotion/drake/blob/32684757f7e91a5599e8eb45ba840375216552f3/drake/systems/plants/RigidBodyTree.h#L767)):*
Although I haven't used the Google style before, I'm finding I like the visual distinction between cheap (accessor) methods and more expensive methods nice.  I think the rule is CamelCase unless the method is so cheap you don't have to think about calling it in a loop. What is confusing about it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

liangfok commented:
*[drake/systems/plants/RigidBodyTree.h, line 767 \[r15\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFQkOEAKnvn4XEzzhcd:-KFWEjALT8vn1lwttuFu:1675632610) ([raw file](https://github.com/robotlocomotion/drake/blob/32684757f7e91a5599e8eb45ba840375216552f3/drake/systems/plants/RigidBodyTree.h#L767)):*
There was confusion above where @amcastro-tri and @david-german-tri did not agree on whether `AddFloatingJoint()` was "cheap". (Note that AddFloatingJoint() is called within the loop that adds models to the rigid body tree). I've submitted a PR  to clarify this rule: https://github.com/RobotLocomotion/drake/pull/2113

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@RussTedrake, since this PR passed the CI test suite and got an LGM from @david-german-tri, I believe it's ready for you to review / merge.

RussTedrake commented:
Reviewed 6 of 14 files at r1, 1 of 8 files at r3, 2 of 7 files at r11, 6 of 6 files at r15.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 1854 \[r15\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFdowkjXt0RH1SRsfu4:-KFdowkjXt0RH1SRsfu5:633604982) ([raw file](https://github.com/robotlocomotion/drake/blob/32684757f7e91a5599e8eb45ba840375216552f3/drake/systems/plants/RigidBodyTree.cpp#L1854)):*
I would prefer AddFloatingJoint  (singular)

---

*[drake/systems/plants/RigidBodyTree.h, line 754 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEraV_vkjMv9DYUNiu:-KFdpjgWnplysBwvWjvX:223867564) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/RigidBodyTree.h#L754)):*
hah.  i disagree.  i think AddFloatingJoint was correct.

---

*[drake/systems/plants/RigidBodyTree.h, line 20 \[r15\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFdpwug9Z_VqSmoR4CQ:-KFdpwug9Z_VqSmoR4CR:-1844993754) ([raw file](https://github.com/robotlocomotion/drake/blob/32684757f7e91a5599e8eb45ba840375216552f3/drake/systems/plants/RigidBodyTree.h#L20)):*
it feels very wrong to have this dependency creep in here (especially since it will be including by every file which includes RigidBodyTree).  it looks like it's for the pose_map only?  

---

*[drake/systems/plants/test/CMakeLists.txt, line 97 \[r15\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFds1k2KLdPhy_rOerW:-KFds1k2KLdPhy_rOerX:1072380374) ([raw file](https://github.com/robotlocomotion/drake/blob/32684757f7e91a5599e8eb45ba840375216552f3/drake/systems/plants/test/CMakeLists.txt#L97)):*
how about just adding this to the CMAKE_CXX_FLAGS in drake/CMakelists.txt iff win32 and GTEST is found?  that would also clean up the other gtest call above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
added my review.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-:-KFdsygNM9h6EQ6nLRmN:1545892743)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@russtedrake, I addressed your comments and am currently waiting on the CI test suite to verify I didn't break anything.

---

Review status: 8 of 19 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 1854 \[r15\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFdowkjXt0RH1SRsfu4:-KFdz5_dMtpVNMZm8caB:-432097505) ([raw file](https://github.com/robotlocomotion/drake/blob/32684757f7e91a5599e8eb45ba840375216552f3/drake/systems/plants/RigidBodyTree.cpp#L1854)):*
Done. Changed back to the singular form: `AddFloatingJoint()`.

---

*[drake/systems/plants/RigidBodyTree.h, line 754 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFEraV_vkjMv9DYUNiu:-KFdtvA6kR9-Ar1RVzzN:-1214826851) ([raw file](https://github.com/robotlocomotion/drake/blob/9932042a3db1fc539fc79ef708242a266aade108/drake/systems/plants/RigidBodyTree.h#L754)):*
I suppose if we use the singular `AddFloatingJoint()`, we could interpret it as "add one floating joint to each of the parent-less links specified in the `link_indices` parameter. Since there is this potential logical explanation, I will revert it back to the singular form.

---

*[drake/systems/plants/RigidBodyTree.h, line 20 \[r15\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFdpwug9Z_VqSmoR4CQ:-KFe-hAznZ1Pgj5jjf6g:-306213417) ([raw file](https://github.com/robotlocomotion/drake/blob/32684757f7e91a5599e8eb45ba840375216552f3/drake/systems/plants/RigidBodyTree.h#L20)):*
Yes, the inclusion of `xmlUtil.h` is only for the definition of type `PoseMap`. I believe the best solution is to put the definitions of both `PoseMap` and `MaterialMap`, which are currently in `xmlUtil.h`, into their own header files. I've updated this PR to include this change.

---

*[drake/systems/plants/test/CMakeLists.txt, line 97 \[r15\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFds1k2KLdPhy_rOerW:-KFePr3EHTDXhim_cwGa:-121756275) ([raw file](https://github.com/robotlocomotion/drake/blob/32684757f7e91a5599e8eb45ba840375216552f3/drake/systems/plants/test/CMakeLists.txt#L97)):*
I've moved this to `[drake distro]/drake/CMakeLists.txt`. Unfortunately, the additional flag in `[drake distro]/CMakeLists.txt` is still required since it is used when compiling the Google Test library, not Drake.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-:-KFeQMAtJiAX1CQ10vVy:592316329)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@david-german-tri I believe I've encountered a CI flake: The OS X CI server went offline during the test: https://drake-jenkins.csail.mit.edu/job/experimental/907/compiler=clang,label=mac/consoleFull

david-german-tri commented:
Yes, that flake is a known issue (e.g. #2039).  @jamiesnape believes it indicates a network hiccup on our colo'ed Mac Minis.  Kitware is having those upgraded those to XServes, which is expected to mitigate the problem.  In the meantime, all you can do is re-run.  :-(

liangfok commented:
@david-german-tri: Thanks for the explanation.

@drake-jenkins-bot retest this please

liangfok commented:
@RussTedrake: This PR passed [CI build 911](https://drake-jenkins.csail.mit.edu/job/experimental/911/) and is thus ready for you to review again.

RussTedrake commented:
Reviewed 9 of 11 files at r16, 2 of 2 files at r17.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/plants/material_map.h, line 2 \[r18\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFfHy_Fr6VYWLjyk95s:-KFfHy_Fr6VYWLjyk95t:-247844018) ([raw file](https://github.com/robotlocomotion/drake/blob/ced0baead823965c552595a1db3f48f92c958bfe/drake/systems/plants/material_map.h#L2)):*
thanks for doing this.  is it worth having an entire header file for this and pose_map?  as opposed to, for instance, putting it at the top of RigidBodyTree.  @david-german-tri , what would you recommend? 

---

*[drake/systems/plants/RigidBodyTree.h, line 756 \[r18\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFfIt8sDaNbOQUE0uOl:-KFfIt8sDaNbOQUE0uOm:-623879567) ([raw file](https://github.com/robotlocomotion/drake/blob/ced0baead823965c552595a1db3f48f92c958bfe/drake/systems/plants/RigidBodyTree.h#L756)):*
document that it's ok for this to be nullptr?  (perhaps it should be later in the argument list, since it's really an optional argument)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/material_map.h, line 2 \[r18\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFfHy_Fr6VYWLjyk95s:-KFfTDEVVDRd596OrWrD:-624129970) ([raw file](https://github.com/robotlocomotion/drake/blob/ced0baead823965c552595a1db3f48f92c958bfe/drake/systems/plants/material_map.h#L2)):*
I think separate headers are the best option for the moment.  If you put `PoseMap` in `RigidBodyTree.h`, you'd then need to include `RigidBodyTree.h` in `xmlUtil.cpp`.

We could remove that dependency by splitting `poseValueToTransform` into an XML-parsing piece and a coordinate-transform piece, and moving the latter out of xmlUtil.  At that point these headers might be superfluous.  But I wouldn't want to cram that extra refactor into this already-complex PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 14 of 19 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/plants/material_map.h, line 2 \[r18\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFfHy_Fr6VYWLjyk95s:-KFfppK8jol1z422sIVO:-1994701324) ([raw file](https://github.com/robotlocomotion/drake/blob/ced0baead823965c552595a1db3f48f92c958bfe/drake/systems/plants/material_map.h#L2)):*
Yes, let's hold off the refactoring of `poseValueToTransform()` since addressing https://github.com/RobotLocomotion/drake/issues/2074 will most likely require refactoring large chunks of XML parsing code out of numerous files including `RigidBodySystem.cpp`, `RigidBodyTree.cpp`, `RigidBodyTreeSDF.cpp` , `RigidBodyTreeURDF.cpp`, and `xmlUtil.cpp`.

---

*[drake/systems/plants/RigidBodyTree.h, line 756 \[r18\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-KFfIt8sDaNbOQUE0uOl:-KFftp6N-7PwIeFOl0Ek:-1296713997) ([raw file](https://github.com/robotlocomotion/drake/blob/ced0baead823965c552595a1db3f48f92c958bfe/drake/systems/plants/RigidBodyTree.h#L756)):*
Done. I moved `pose_map` to be the last parameter of method `AddFloatingJoint()`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@RussTedrake, this PR passed [CI build 917](https://drake-jenkins.csail.mit.edu/job/experimental/917/) and is thus ready for another round of review.

RussTedrake commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r19.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2053#-:-KFgwgbTORGeVoP4iyHw:1417088721)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Somehow, this has added 30 minutes to the linux-gcc-continuous-debug compile time.

jwnimmer-tri commented:
Well, perhaps stated more carefully: the linux-gcc-continuous-debug compile time increased by 45 minutes when this PR was merged, and has not gotten better since.  Possibly it's a CI problem that surfaced coincidentally.  I'm still looking.

liangfok commented:
Interesting. Assuming my code changes is not the culprit (a **big** assumption, I know...), one possible explanation is our CI processes are running on a shared server and someone else's process was moved onto the same server consuming a huge amount of computational resources.
